compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -1606,9 +1607,18 @@ class TypeComparer(initctx: Context) extends ConstraintHandling[AbsentContext] w
         formals2 match {
           case formal2 :: rest2 =>
             val formal2a = if (tp2.isParamDependent) formal2.subst(tp2, tp1) else formal2
+            // The next two definitions handle the special case mentioned above, where
+            // the Java argument has type 'Any', and the Scala argument has type 'Object' or
+            // 'Object|Null', depending on whether explicit nulls are enabled.
+            lazy val formal2IsObject =
+              if (ctx.explicitNulls) formal2.isNullableUnion && formal2.stripNull(ctx).isRef(ObjectClass)
```

Comment:
It's better to make this (and the next lazy val a `def`. Local lazy vals allocate closures and are therefore expensive even if they are not used.

Also, I did not see what is gained here by the `isNullableUnion` test?

Reply:
Changed to `def`; also use `OrNull` extractor to avoid normalizing twice.